### PR TITLE
Convert ObjectLiteralExpr to not use tc.callWitness() or generate a SemanticExpr

### PR DESCRIPTION
**Pull request by slavapestov**
This takes the patch from https://github.com/apple/swift/pull/23624 and adds a fix for a tricky bug in ProtocolConformance::getWitnessDeclRef(). Also a couple of tiny related cleanups. Thanks @pschuh for the actual refactoring!

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test

File: lib/AST/ProtocolConformance.cpp [hunk 1]
```diff
@@ -966,8 +966,23 @@ Witness SelfProtocolConformance::getWitness(ValueDecl *requirement,
...
 RootProtocolConformance::getWitnessDeclRef(ValueDecl *requirement,
                                            LazyResolver *resolver) const {
-  if (auto witness = getWitness(requirement, resolver))
-    return witness.getDeclRef();
+  if (auto witness = getWitness(requirement, resolver)) {
+    auto *witnessDecl = witness.getDecl();
+
+    // If the witness is generic, you have to call getWitness() and build
+    // your own substitutions in terms of the synthetic environment.
+    if (auto *witnessDC = dyn_cast<DeclContext>(witnessDecl))
```
Comment (pschuh) [inline lib/AST/ProtocolConformance.cpp]: Would this be triggered by just another level of genericness in the crasher example?

**Review by pschuh — APPROVED** [review]

File: lib/AST/ProtocolConformance.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/AST/ProtocolConformance.cpp]: I’ll take a look. It would not have worked before anyway though.

**Comment by compnerd** [comment]
This broke the windows tests!

**Comment by compnerd** [comment]
I would have expected this to break 32-bit iOS as well

**Pull request merged**